library/Booster/Pattern/ApplyEquations.hs

```diff
@@ -391,34 +391,25 @@ iterateEquations direction preference startTerm = do
             -- the equation evaluation does not change the term any more.
             resetChanged
             -- evaluate functions and simplify (recursively at each level)
-            newTerm <- applyTerm direction preference llvmResult
+            newTerm <-
+                let simp = cached Equations $ traverseTerm direction simp (applyHooksAndEquations preference)
+                 in simp llvmResult
             changeFlag <- getChanged
             if changeFlag
                 then checkForLoop newTerm >> resetChanged >> go newTerm
                 else pure llvmResult
 
 llvmSimplify :: forall io. MonadLoggerIO io => Term -> EquationT io Term
-llvmSimplify t = do
+llvmSimplify term = do
     config <- getConfig
     case config.llvmApi of
-        Nothing -> pure t
+        Nothing -> pure term
         Just api -> do
             logOtherNS "booster" (LevelOther "Simplify") "Calling LLVM simplification"
-            llvmEval config.definition api t
-
-llvmEval :: MonadLoggerIO io => KoreDefinition -> LLVM.API -> Term -> EquationT io Term
-llvmEval definition api = eval
+            let simp = cached LLVM $ evalLlvm config.definition api $ traverseTerm BottomUp simp pure
+             in simp term
   where
-    eval t@(Term attributes _)
-        | attributes.isEvaluated = pure t
-        | otherwise =
-            fromCache LLVM t >>= \case
-                Nothing ->
-                    eval' t
-                Just cached ->
-                    when (cached /= t) setChanged >> pure cached
-
-    eval' t@(Term attributes _)
+    evalLlvm definition api cb t@(Term attributes _)
```

Comment:
I think this function deserves to have:
* an explicit type signature
* a better name for the `cb` parameter, together with the comment on what the callback does. How about we call it `traversalAction` or something along these lines?